File: code/234-Palindrome Linked List/234-Palindrome Linked List.c
/*
LeetCode 234 - Palindrome Linked List

解题步骤（自动整理）
1. 用快慢指针/虚拟头节点等技巧处理链表边界
2. 按题意进行遍历、反转、合并或断链等操作
3. 返回处理后的链表头节点或计算结果
*/
// Given the head of a singly linked list, return true if it is a palindrome.



// Example 1:


// Input: head = [1,2,2,1]
// Output: true
// Example 2:


// Input: head = [1,2]
// Output: false


// Constraints:

// The number of nodes in the list is in the range [1, 10^5].
// 0 <= Node.val <= 9


// Follow up: Could you do it in O(n) time and O(1) space?

bool isPalindrome(struct ListNode* head){
    int *q = malloc(sizeof(int) * 100001);
    struct ListNode *p = head;
    int l = 0, r = 0;
    while (p) {
        q[r ++] = p->val;
        p = p->next;
    }
    r --;
    while (l < r) {
        if (q[l ++] != q[r --]) {
            return false;
        }
    }
    return true;
}

File: code/234-Palindrome Linked List/234-Palindrome Linked List.py
"""
LeetCode 234 - Palindrome Linked List

解题步骤（自动整理）
1. 将起点/初始层入队，并用 visited 或原地标记避免重复访问
2. 按层或按队列弹出元素扩展相邻节点，满足条件时更新答案
3. 队列处理完毕后返回结果（或在首次到达目标时提前返回）
"""
# Given a singly linked list, determine if it is a palindrome.
#
# Example 1:
# Input: 1->2
# Output: false

# Example 2:
# Input: 1->2->2->1
# Output: true
# Follow up:
# Could you do it in O(n) time and O(1) space?

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        deque = []
        curr = head
        while curr is not None:
            deque.append(curr.val)
            curr = curr.next
        first, second = 0, len(deque) - 1
        while first < second:
            if deque[first] != deque[second]:
                return False
            else:
                first += 1
                second -= 1
        return True
